File: AlgorithmsPackJavaSolutions/Detect Cycles in 2D Grid.java
class Solution {
    private final int[] dr = {-1, 0, 1, 0};
    private final int[] dc = {0, 1, 0, -1};
    private boolean isCycle;

    private boolean isValid(int r, int c, char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        return r >= 0 && r < rows && c >= 0 && c < cols;
    }

    private void dfs(int r, int c, char[][] grid, boolean[][] visited, char oldColor, int parentRow, int parentCol) {
        if (!isValid(r, c, grid) || grid[r][c] != oldColor || isCycle) {
            return;
        }

        if (visited[r][c]) {
            isCycle = true;
            return;
        }

        visited[r][c] = true;

        for (int d = 0; d < 4; d++) {
            int nr = dr[d] + r;
            int nc = dc[d] + c;

            if (nr == parentRow && nc == parentCol) {
                continue;
            }

            dfs(nr, nc, grid, visited, oldColor, r, c);
        }
    }

    public boolean containsCycle(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        boolean[][] visited = new boolean[rows][cols];

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (!visited[r][c]) {
                    isCycle = false;
                    dfs(r, c, grid, visited, grid[r][c], -1, -1);
                    if (isCycle) {
                        return true;
                    }
                }
            }
        }

        return false;
    }
}

File: AlgorithmsPackCPP/Detect Cycles in 2D Grid.cpp
class Solution {
int dr[4]{-1,0,1,0};
int dc[4]{0,1,0,-1};
bool is_cycle;  
public:
bool isValid(int r, int c, vector<vector<char>>&grid){
    if(r < 0 || r >= grid.size())
        return false;
    if(c<0 || c>=grid[0].size())
        return false;
    return true;
}
/*
parent_row and parent_col are the 'node we come from', 
and we need them in this solution
*/
void dfs(int r, int c, vector<vector<char>>&grid, vector<vector<bool>>&visited, char oldColor, int parent_row = -1, int parent_col = -1) 
{
    if(!isValid(r,c,grid) || grid[r][c]!=oldColor||is_cycle)
        return;
    
    if(visited[r][c]){
        is_cycle = true;
        return;
    }
    
    visited[r][c] = 1;
    
    for(int d = 0; d < 4; d++){   
        int nr = dr[d]+r;
        int nc = dc[d]+c;
         
        //THIS IS THE PARENT 'TRICK'.  We can't 'go back' to our parent.
        if(nr==parent_row && nc==parent_col)
            continue;
        
        dfs(nr,nc,grid,visited,oldColor,r,c);
    }
}

bool containsCycle(vector<vector<char>>& grid) {
    vector<vector<bool>>visited(grid.size(),vector<bool>(grid[0].size()));
    int rows = grid.size();
    int cols = grid[0].size();        
    for(int r = 0; r < rows; r++){
        for(int c = 0; c < cols; c++){
            if(!visited[r][c]){ 
                is_cycle = false; 
                dfs(r,c,grid,visited,grid[r][c]);
                if(is_cycle)
                    return true;
            }
        }
    }
    return false;
};
